File: Chapter 02/C#/2.3.4 Performance/DashboardAfter.cs
namespace Chapter02.Performance;

/// <summary>
/// Represents a dashboard service that uses a cache to improve performance.
/// This class demonstrates the "Cache-Aside" pattern.
/// </summary>
public class DashboardAfter
{
    // Using a constant for the cache's Time-To-Live (TTL) is a good practice.
    // It makes the code more readable and ensures the expiration policy is consistent.
    private const int CACHE_TTL_SECONDS = 600; // 10 minutes
    
    private readonly DatabaseService _databaseService = new();
    private readonly CacheService _cache = new();

    /// <summary>
    /// Gets a summary of dashboard data for a user, using a cache to optimize performance.
    /// </summary>
    /// <param name="userId">The ID of the user.</param>
    /// <returns>An object containing the user's dashboard data.</returns>
    public object GetDashboardSummary(string userId)
    {
        string cacheKey = $"dashboard:{userId}";

        // IMPROVEMENT: The "Cache-Aside" Pattern
        //
        // STEP 1: Check the cache first.
        // Before doing any expensive work, we check if the data we need is already
        // in our fast in-memory cache. A cache read is significantly faster (e.g., <5ms)
        // than a database query (e.g., 500ms).
        var cachedDashboard = _cache.Get(cacheKey);
        
        // If `cachedDashboard` is not null, we have a "cache hit."
        // We can immediately return the cached data without touching the database.
        if (cachedDashboard != null)
        {
            return cachedDashboard;
        }

        // STEP 2: Handle a "cache miss."
        // If the data is not in the cache, we proceed with the expensive operation:
        // fetching the data from the database.
        var profile = _databaseService.GetProfile(userId);
        var orders = _databaseService.GetOrders(userId);
        var activity = _databaseService.GetActivity(userId);

        var dashboardData = new { profile, orders, activity };

        // STEP 3: Store the result in the cache.
        // Before returning the data, we save it to the cache. The next time this
        // method is called for the same user (within the TTL window), we'll get a
        // cache hit and avoid the database calls altogether.
        _cache.Set(cacheKey, dashboardData, CACHE_TTL_SECONDS);

        return dashboardData;
    }
}

File: Chapter 02/JS/2.3.4 Performance/dashboardAfter.js
const databaseService = require('./databaseService');
const cache = require('./cacheService');

// Using a constant for the cache's Time-To-Live (TTL) is a good practice.
const CACHE_TTL_SECONDS = 600; // 10 minutes

/**
 * Gets a summary of dashboard data for a user, using a cache to optimize performance.
 * This function demonstrates the "Cache-Aside" pattern.
 * @param {string} userId - The ID of the user.
 * @returns {object} An object containing the user's dashboard data.
 */
function getDashboardSummaryAfter(userId) {
    const cacheKey = `dashboard:${userId}`;
    
    // IMPROVEMENT: The "Cache-Aside" Pattern
    //
    // STEP 1: Check the cache first.
    // Before doing expensive work, we check if the data is in our fast in-memory cache.
    const cachedDashboard = cache.get(cacheKey);
    
    // If we get a "cache hit," we can immediately return the cached data.
    if (cachedDashboard) {
        return cachedDashboard;
    }
    
    // STEP 2: Handle a "cache miss."
    // If the data is not in the cache, we proceed with the expensive operation.
    const profile = databaseService.getProfile(userId);
    const orders = databaseService.getOrders(userId);
    const activity = databaseService.getActivity(userId);
    
    const dashboardData = { profile, orders, activity };
    
    // STEP 3: Store the result in the cache.
    // Before returning, we save the data to the cache. The next request for this
    // user will get a cache hit.
    cache.set(cacheKey, dashboardData, CACHE_TTL_SECONDS);
    
    return dashboardData;
}

module.exports = getDashboardSummaryAfter;